### PR TITLE
Technical guidelines: Clarify what to do when service contracts already exist in a module not suffixed with "Api"

### PR DESCRIPTION
**Pull request by nathanjosiah**
## This PR is a:

- [ ] New topic
- [ ] Content update
- [ ] Content fix or rewrite
- [X] Bug fix or improvement

## Summary

Re: https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html#64-service-contracts-application-layer

For the new service contract doc requirements, what should be done when new interfaces need to be added to a module that already has the contracts in the main module. For example, I need to add contracts to support `MyModule` but `MyModule/Api` already exists. Should they be added to `MyModule/Api`? or should the contract locations be fragmented and create the new module `MyModuleApi/Api/` with the new contracts and leave the `MyModule/Api` contracts as-is?

## More information regarding suggested change

My suggestion is to keep the contracts in the module rather than making a new module and deprecating the old contracts. This will prevent 3rd party devs from deprecating code and it also will simplify the development process for module that already have contracts. These modules can be migrated to the new form at a later time with specific direction when it makes sense in the project lifecycles (i.e. a major or minor release).

whatsnew
Added clarification about the location of Service Contracts in section 6.4.1.1 of the [Technical Guidelines](https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/devdocs?pullRequest=3524) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/devdocs?pullRequest=3524) before we can accept your contribution.<br/>

**Comment by magento-engcom-team** [comment]
@nathanjosiah thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by magento-cicd2** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.2/coding-standards/technical-guidelines.md [hunk 1]
```diff
@@ -536,7 +536,7 @@ class View extends Template
 
 6.4.1.1. Service contract interfaces SHOULD be placed in separate API modules. The other modules will depend on the API module, and implementations could be easily swapped via `di.xml`. API module namess must end with the `Api` suffix. For example, if a module is named `MyModule`, its APIs SHOULD be declared in a module named `MyModuleApi`.
 
-6.4.1.2. Service interfaces that should be exposed as web APIs MUST be placed under the `MyModuleApi/Api` directory. Service data interfaces MUST be placed under `MyModuleApi/Api/Data`. Directories under `MyModuleApi/Api` SHOULD NOT be nested.
+6.4.1.2. Service interfaces that should be exposed as web APIs MUST be placed under the `MyModuleApi/Api` directory. Service data interfaces MUST be placed under `MyModuleApi/Api/Data`. Directories under `MyModuleApi/Api` SHOULD NOT be nested. If contracts need to be added to support a module that already contains classes under `Api/` (e.g. `MyModule/Api`), the contracts should be added to the existing module.
```
Comment (paliarush) [inline guides/v2.2/coding-standards/technical-guidelines.md]:
```suggestion
6.4.1.2. Service interfaces that should be exposed as web APIs MUST be placed under the `MyModuleApi/Api` directory. Service data interfaces MUST be placed under `MyModuleApi/Api/Data`. Directories under `MyModuleApi/Api` SHOULD NOT be nested. If contracts need to be added to support a module that already contains classes under `Api/` (e.g. `MyModule/Api`), the contracts SHOULD be added to the existing module.  
```

**Comment by navarr** [comment]
Doesn't this burden eventually moving existing contracts into an Api module?

Wouldn't it be better to `@deprecate` the old ones, place them in a new service contract module, have the new interfaces extend the old interfaces, and give plenty of time for implementations to switch?

**Comment by keharper** [comment]
@buskamuza Is any additional architect review required?

**Comment by buskamuza** [comment]
Please don't merge yet.

**Comment by keharper** [comment]
@buskamuza You closed https://github.com/magento/architecture/issues/75 last week. Can this PR be merged?

File: guides/v2.2/coding-standards/technical-guidelines.md [hunk 2]
```diff
@@ -536,7 +536,7 @@ class View extends Template
 
 6.4.1.1. Service contract interfaces SHOULD be placed in separate API modules. The other modules will depend on the API module, and implementations could be easily swapped via `di.xml`. API module namess must end with the `Api` suffix. For example, if a module is named `MyModule`, its APIs SHOULD be declared in a module named `MyModuleApi`.
 
-6.4.1.2. Service interfaces that should be exposed as web APIs MUST be placed under the `MyModuleApi/Api` directory. Service data interfaces MUST be placed under `MyModuleApi/Api/Data`. Directories under `MyModuleApi/Api` SHOULD NOT be nested.
+6.4.1.2. Service interfaces that should be exposed as web APIs MUST be placed under the `MyModuleApi/Api` directory. Service data interfaces MUST be placed under `MyModuleApi/Api/Data`. Directories under `MyModuleApi/Api` SHOULD NOT be nested. If contracts need to be added to support a module that already contains classes under `Api/` (e.g. `MyModule/Api`), the contracts SHOULD be added to the existing module.
```
Comment (buskamuza) [inline guides/v2.2/coding-standards/technical-guidelines.md]: I propose to add a note to 6.4.1.1. instead as that is the item describing the rules on a module level. So I'd rather rephrase that one as "6.4.1.1. Service contract interfaces SHOULD be placed in separate API modules **, except when existing module already contains Service Contracts in `Api` folder**. The other modules will depend on the API module, and implementations could be easily swapped via `di.xml`. API module namess must end with the `Api` suffix. For example, if a module is named `MyModule`, its APIs SHOULD be declared in a module named `MyModuleApi`."

An important point here is that `Api` folder should contain **Service Contracts** (probably that's the only expected, but who knows).

**Comment by buskamuza** [comment]
@keharper , we close tickets just to indicate that it's a past meeting.

**Comment by keharper** [comment]
running tests

**Comment by keharper** [comment]
running tests

**Pull request merged**